File: docs/js/state.js
import { SoundFX, Camera, Player } from './classes.js';
import { CONFIG, CONTROLS_P1, CONTROLS_P2, TIMING, COLORS, DIFFICULTIES } from './config.js';
import { DIFFICULTY_PRESETS } from './ai/difficulty.js';

let _state = {
    frameCount: 0,
    maze: [],
    players: [],
    mines: [],
    particles: [],
    portals: [],
    projectiles: [],
    ammoCrate: null,
    ammoLastTakeTime: 0,
    keys: {},
    gameTime: 0,
    maxGameTime: 0,
    isGameOver: false,
    isRoundOver: false,
    deathTimer: 0,
    victimIdx: -1,
    isPaused: false,
    isDraw: false,
    onlineTransitionPending: false, // Prevents double round/game transitions in online mode
    messages: {
        deathReason: "",
        win: "",
        taunt: "",
        round: "",
        winColor: "#fff",
        roundColor: "#fff"
    },
    scrollX: 70,
    scrollY: 0,
    scrollXVal: -1,
    scrollYVal: +2,
    sfx: new SoundFX(),
    camera: new Camera(),
    gpData: null,
    portalReverseColors: false,
    cyanColor: COLORS.find(x => x.name === "CYAN").hex,
    blueColor: COLORS.find(x => x.name === "BLUE").hex,
    highScores: (() => {
        try {
            return JSON.parse(localStorage.getItem(CONFIG.STORAGE_KEY)) || [];
        } catch (e) {
            console.warn('Failed to load high scores:', e);
            return [];
        }
    })() || [
        { name: "ZEU", winColor: "#aa00ffff", oppColor: "#ff0000ff", score: 10, oppScore: 6, opponent: "CPU-INSANE", multiplier: 1 },
        { name: "ARE", winColor: "#ffffffff", oppColor: "#ff5100ff", score: 8, oppScore: 5, opponent: "CPU-HARD", multiplier: 0.8 },
        { name: "HER", winColor: "#00aaffff", oppColor: "#ffff00ff", score: 8, oppScore: 5, opponent: "CPU-INTERME", multiplier: 0.4 }
    ],
    // Match statistics
    stats: (() => {
        try {
            return JSON.parse(localStorage.getItem(CONFIG.STORAGE_KEY + '_stats')) || {
                totalMatches: 0,
                wins: 0,
                losses: 0,
                draws: 0,
                totalRounds: 0,
                totalPlayTime: 0, // in frames
                byDifficulty: {}, // { "HARD": { wins: 0, losses: 0 }, ... }
                byMode: { SINGLE: { wins: 0, losses: 0 }, MULTI: { wins: 0, losses: 0 }, ONLINE: { wins: 0, losses: 0 } },
                recentMatches: [] // Last 10 matches
            };
        } catch (e) {
            console.warn('Failed to load stats:', e);
            return {
                totalMatches: 0, wins: 0, losses: 0, draws: 0,
                totalRounds: 0, totalPlayTime: 0,
                byDifficulty: {},
                byMode: { SINGLE: { wins: 0, losses: 0 }, MULTI: { wins: 0, losses: 0 }, ONLINE: { wins: 0, losses: 0 } },
                recentMatches: []
            };
        }
    })(),
    matchStartFrame: 0, // Track when match started for duration calculation
    pauseMenuSelection: 0, // 0: Resume, 1: Restart, 2: Quit
    highScoreTab: 0, // 0: Leaderboard, 1: Stats
    playerSetup: {
        activePlayer: 0,      // 0 for P1, 1 for P2
        difficultyIdx: 3,     // Which difficulty selected
        colorIdx: 0,          // Which color selected
        nameCharIdx: 0,       // Which character in name (0, 1, 2)
        nameChars: [65, 65, 65],  // ASCII codes
        phase: 'DIFFICULTY',       // 'DIFFICULTY', 'COLOR' or 'NAME'
        isDone: false
    },
    difficulty: 'INSANE',
    aiMentalModel: {
        strategy: null,
        moveDir: { dx: 0, dy: 0 },
        energyStrat: { shield: false, boost: false },
        lastThinkTime: 0
    },
    gameMode: 'SINGLE', // 'SINGLE', 'MULTI', 'ONLINE', 'HIGHSCORES'
    screen: 'MENU',
    isAttractMode: false,
    demoResetTimer: 0,
    inputDelay: CONFIG.INPUT_DELAY,
    menuSelection: 0, // 0: SINGLE, 1: LOCAL MULTI, 2: ONLINE MULTI, 3: HIGH SCORES
    lastUpdateTime: 0,
    accumulator: 0,
};

export function getState() {
    return _state;
}

export function updateState(updater) {
    const updates = typeof updater === 'function' ? updater(_state) : updater;
    _state = { ..._state, ...updates };
}

export function saveHighScore() {
    const state = getState();
    let victimIdx = state.victimIdx;
    let winnerIdx = (victimIdx === 0) ? 1 : 0;
    let winner = state.players[winnerIdx];
    let opponent = state.players[victimIdx];
    const ps = state.playerSetup;
    const diff = DIFFICULTY_PRESETS[DIFFICULTIES[ps.difficultyIdx].name];
    let victimName = opponent.name === "CPU" ? `CPU-${diff.NAME.substring(0, 7)}` : `${opponent.name}`;
    let highScores = JSON.parse(JSON.stringify(state.highScores)); // Deep copy
    let entry = highScores.find(e => e.name === winner.name && e.opponent === victimName && e.multiplier === diff.HIGHSCORE_MULTIPLIER);
    let oppColor = opponent.name === "CPU" ? diff.COLOR : opponent.color;
    if (entry) {
        entry.score += winner.score;
        entry.oppScore += opponent.score;
    } else {
        highScores.push({
            name: winner.name,
            winColor: winner.color,
            oppColor: oppColor,
            score: winner.score,
            oppScore: opponent.score,
            opponent: victimName,
            multiplier: diff.HIGHSCORE_MULTIPLIER
        });
    }
    // Sort by wins (descending) and keep top 5
    highScores.sort((a, b) => ((b.score - b.oppScore) * b.multiplier) - ((a.score - a.oppScore) * a.multiplier));
    highScores = highScores.slice(0, 10);

    updateState({ highScores });
    localStorage.setItem(CONFIG.STORAGE_KEY, JSON.stringify(highScores));
}

/**
 * Save match statistics to localStorage
 */
export function saveStats() {
    try {
        localStorage.setItem(CONFIG.STORAGE_KEY + '_stats', JSON.stringify(getState().stats));
    } catch (e) {
        console.warn('Failed to save stats:', e);
    }
}

/**
 * Record match result in statistics
 * @param {number} winnerIdx - Index of winner (0 or 1), or -1 for draw
 */
export function recordMatchStats(winnerIdx) {
    const state = getState();
    const stats = JSON.parse(JSON.stringify(state.stats));

    const p1 = state.players[0];
    const p2 = state.players[1];
    const isP1Human = p1.name !== "CPU";
    const isP2Human = p2.name !== "CPU";
    const ps = state.playerSetup;
    const diffName = DIFFICULTIES[ps.difficultyIdx]?.name || 'UNKNOWN';
    const mode = getState().gameMode;
    const matchDuration = state.frameCount - state.matchStartFrame;

    // Update totals
    stats.totalMatches++;
    stats.totalPlayTime += matchDuration;
    stats.totalRounds += Math.max(p1.score, p2.score);

    // Determine win/loss from human player perspective
    let result = 'draw';
    if (winnerIdx === 0 && isP1Human) result = 'win';
    else if (winnerIdx === 1 && isP1Human && !isP2Human) result = 'loss';
    else if (winnerIdx === 1 && isP2Human) result = 'win';
    else if (winnerIdx === 0 && isP2Human && !isP1Human) result = 'loss';
    else if (winnerIdx === -1) result = 'draw';

    // Update win/loss/draw counts
    if (result === 'win') stats.wins++;
    else if (result === 'loss') stats.losses++;
    else stats.draws++;

    // Update by difficulty (for single player)
    if (mode === 'SINGLE') {
        if (!stats.byDifficulty[diffName]) {
            stats.byDifficulty[diffName] = { wins: 0, losses: 0 };
        }
        if (result === 'win') stats.byDifficulty[diffName].wins++;
        else if (result === 'loss') stats.byDifficulty[diffName].losses++;
    }

    // Update by mode
    if (!stats.byMode[mode]) {
        stats.byMode[mode] = { wins: 0, losses: 0 };
    }
    if (result === 'win') stats.byMode[mode].wins++;
    else if (result === 'loss') stats.byMode[mode].losses++;

    // Add to recent matches (keep last 10)
    stats.recentMatches.unshift({
        date: Date.now(),
        mode: mode,
        difficulty: mode === 'SINGLE' ? diffName : null,
        p1Name: p1.name,
        p1Score: p1.score,
        p2Name: p2.name,
        p2Score: p2.score,
        winner: winnerIdx === 0 ? p1.name : (winnerIdx === 1 ? p2.name : 'DRAW'),
        duration: Math.round(matchDuration / 60) // in seconds
    });
    stats.recentMatches = stats.recentMatches.slice(0, 10);

    updateState({ stats });
    saveStats();
}

/**
 * Get formatted stats for display
 */
export function getFormattedStats() {
    const stats = getState().stats;
    const winRate = stats.totalMatches > 0
        ? Math.round((stats.wins / stats.totalMatches) * 100)
        : 0;
    const avgDuration = stats.totalMatches > 0
        ? Math.round(stats.totalPlayTime / stats.totalMatches / 60)
        : 0;

    return {
        totalMatches: stats.totalMatches,
        wins: stats.wins,
        losses: stats.losses,
        draws: stats.draws,
        winRate: winRate,
        avgDuration: avgDuration,
        totalPlayTime: Math.round(stats.totalPlayTime / 3600), // in minutes
        byDifficulty: stats.byDifficulty,
        byMode: stats.byMode,
        recentMatches: stats.recentMatches
    };
}

export function suddenDeathIsActive() {
    const state = getState();
    if (state.gameTime)
        return state.gameTime <= TIMING.SUDDEN_DEATH_TIME;
    else return false;
}

export function shouldSpawnAmmoCrate() {
    const state = getState();
    if (state.gameTime && !state.ammoCrate)
        return state.frameCount - state.ammoLastTakeTime > TIMING.AMMO_RESPAWN_DELAY;
    else return false;
}

export function resetStateForMatch() {
    const state = getState();
    // Store current names if they exist
    let CPUColors = COLORS.filter(x => x.name !== 'BLACK' && x.name !== 'ORANGE' && x.name !== 'BLUE' && x.name !== 'RED' && x.name !== 'PURPLE')
    let p1Name = state.players[0]?.name || "CPU";
    let p1Color = state.players[0]?.color ?? CPUColors[Math.floor(Math.random() * CPUColors.length)]?.hex;
    let p2Name = state.gameMode === 'MULTI' ? state.players[1]?.name || "CPU" : "CPU";
    CPUColors = CPUColors.filter(x => x.hex !== p1Color);
    let randomColor2 = CPUColors[Math.floor(Math.random() * CPUColors.length)]?.hex
    let p2Color = state.gameMode === 'MULTI' ? (state.players[1]?.color ?? randomColor2) : randomColor2;

    const newPlayers = [
        new Player(0, p1Color, CONTROLS_P1),
        new Player(1, p2Color, CONTROLS_P2)
    ];

    newPlayers[0].name = p1Name;
    newPlayers[1].name = p2Name;

    updateState({
        players: newPlayers,
        frameCount: 0,
        isGameOver: false,
        isRoundOver: false,
        maze: [],
        mines: [],
        particles: [],
        portals: [],
        projectiles: [],
        ammoCrate: null,
        gameTime: CONFIG.GAME_TIME,
        maxGameTime: CONFIG.GAME_TIME,
        deathTimer: 0,
        victimIdx: -1,
        isPaused: false,
        isDraw: false,
        onlineTransitionPending: false,
        matchStartFrame: 0,
        pauseMenuSelection: 0,
        messages: {
            deathReason: "",
            win: "",
            taunt: "",
            round: "",
            winColor: "#fff",
            roundColor: "#fff"
        },
        scrollX: 70,
        scrollY: 0,
        scrollXVal: -1,
        scrollYVal: +2,
        portalReverseColors: false
    });
}

File: pi/state.py
# pi/state.py
# Central game state management ported from docs/js/state.js

import os
import json
import random

from config import (
    COLORS, TIMING, GAME_TIME, INPUT_DELAY,
)
from classes import Camera, Player

HIGHSCORE_PATH = os.path.join(os.path.dirname(__file__), 'highscores.json')

_DEFAULT_HIGH_SCORES = [
    {'name': 'ZEU', 'win_color': '#aa00ffff', 'opp_color': '#ff0000ff', 'score': 10, 'opp_score': 6, 'opponent': 'CPU-INSANE', 'multiplier': 1},
    {'name': 'ARE', 'win_color': '#ffffffff', 'opp_color': '#ff5100ff', 'score': 8, 'opp_score': 5, 'opponent': 'CPU-HARD', 'multiplier': 0.8},
    {'name': 'HER', 'win_color': '#00aaffff', 'opp_color': '#ffff00ff', 'score': 8, 'opp_score': 5, 'opponent': 'CPU-INTERME', 'multiplier': 0.4},
]


def _load_high_scores():
    """Load high scores from JSON file, returning defaults on failure."""
    try:
        if os.path.exists(HIGHSCORE_PATH):
            with open(HIGHSCORE_PATH, 'r', encoding='utf-8') as f:
                data = json.load(f)
                if isinstance(data, list):
                    return data
    except Exception as e:
        print(f'[state] Failed to load high scores: {e}')
    return list(_DEFAULT_HIGH_SCORES)


def save_high_scores(state):
    """Persist high scores to JSON file."""
    try:
        with open(HIGHSCORE_PATH, 'w', encoding='utf-8') as f:
            json.dump(state.high_scores, f, indent=2)
    except Exception as e:
        print(f'[state] Failed to save high scores: {e}')


def sudden_death_is_active(state):
    """Return True if game time is at or below the sudden-death threshold."""
    if state.game_time:
        return state.game_time <= TIMING.SUDDEN_DEATH_TIME
    return False


def should_spawn_ammo_crate(state):
    """Return True if it's time to spawn a new ammo crate."""
    if state.game_time and not state.ammo_crate:
        return state.frame_count - state.ammo_last_take_time >= TIMING.AMMO_RESPAWN_DELAY
    return False


def get_two_player_colors():
    """Return two random distinct, non-BLACK, non-WHITE Color objects."""
    excluded = {'BLACK', 'WHITE'}
    available = [c for c in COLORS if c.name not in excluded]
    p1_color = random.choice(available)
    rest = [c for c in available if c is not p1_color]
    p2_color = random.choice(rest)
    return [p1_color, p2_color]


def reset_state_for_match(state, p1_color_idx=None, p2_color_idx=None):
    """Reset game state for a new match, preserving player names and scores."""
    excluded = {'BLACK', 'ORANGE', 'BLUE', 'RED', 'PURPLE'}
    available = [c for c in COLORS if c.name not in excluded]

    # Determine P1 colour
    if p1_color_idx is not None:
        p1_color = COLORS[p1_color_idx]
    elif state.players[0] and state.players[0].color:
        p1_color = state.players[0].color
    else:
        p1_color = random.choice(available)

    avail2 = [c for c in available if c is not p1_color]

    # Determine P2 colour
    if p2_color_idx is not None:
        p2_color = COLORS[p2_color_idx]
    elif state.players[1] and state.players[1].color:
        p2_color = state.players[1].color
    else:
        p2_color = random.choice(avail2) if avail2 else available[0]

    # Preserve names
    p1_name = state.players[0].name if state.players[0] else 'CPU'
    p2_name = state.players[1].name if (state.game_mode == 'MULTI' and state.players[1]) else 'CPU'

    new_p1 = Player(0, name=p1_name, color=p1_color)
    new_p2 = Player(1, name=p2_name, color=p2_color)

    state.players = [new_p1, new_p2]
    state.frame_count = 0
    state.is_game_over = False
    state.is_round_over = False
    state.maze = []
    state.mines = []
    state.particles = []
    state.portals = []
    state.projectiles = []
    state.ammo_crate = None
    state.ammo_last_take_time = -999
    state.game_time = GAME_TIME
    state.max_game_time = GAME_TIME
    state.death_timer = 0
    state.victim_idx = -1
    state.is_paused = False
    state.is_draw = False
    state.pause_menu_selection = 0
    state.portal_reverse_colors = False
    state.scroll_x = 0
    state.messages = {
        'death_reason': '',
        'win': '',
        'taunt': '',
        'round': '',
        'win_color': None,
        'round_color': None,
    }


class GameState:
    """Central game state object."""

    def __init__(self):
        # Game timing
        self.frame_count = 0
        self.game_time = 0
        self.max_game_time = 0

        # Entities
        self.maze = []
        self.players = [None, None]
        self.mines = []
        self.particles = []
        self.portals = []
        self.projectiles = []
        self.ammo_crate = None
        self.ammo_last_take_time = -999

        # Game flow
        self.is_game_over = False
        self.is_round_over = False
        self.death_timer = 0
        self.victim_idx = -1
        self.is_paused = False
        self.is_draw = False

        # Messages
        self.messages = {
            'death_reason': '',
            'win': '',
            'taunt': '',
            'round': '',
            'win_color': None,
            'round_color': None,
        }

        # Visual state
        self.scroll_x = 0
        self.camera = Camera()
        self.portal_reverse_colors = False

        # High scores (loaded from JSON file)
        self.high_scores = _load_high_scores()

        # Screen state
        self.screen = 'MENU'
        self.game_mode = 'SINGLE'
        self.is_attract_mode = False
        self.demo_reset_timer = 0
        self.difficulty = 'INTERMEDIATE'
        self.menu_selection = 0
        self.pause_menu_selection = 0
        self.high_score_tab = 0
        self.input_delay = INPUT_DELAY

        # Player setup
        self.player_setup = {
            'active_player': 0,
            'difficulty_idx': 1,
            'color_idx': 0,
            'name_char_idx': 0,
            'name_chars': [ord('A'), ord('A'), ord('A')],
            'phase': 'DIFFICULTY',
            'is_done': False,
        }

        # Input state
        self.gamepad_state = {0: {}, 1: {}}
        self.keyboard_state = {}
        self.running = True
